README.md
# 🥱 react-lazy-io

## 📖 Preface

Recently at my
[work](https://twitter.com/imshubhamsingh_/status/1437468156616855553), I was
trying to implement a simple lazy load. I ended up creating a simple, functional
component that uses Intersection Observer (Or Scroll Event for older browser) to
determine the visibility of images. In my original, implementation all the
elements were creating their Intersection Observer, which is fine as long as
there are limited elements. But to take it to the next level, a question arises:
What if I can use a single listener or observer for my Lazy loaded element. This
library is a simple approach toward using context to share a single observer
instance and using a lookup functionally, the Lazy items can be added and
deleted from the observer.

## 🧑🏻‍💻 Install

```
npm install react-lazy-io
```

or

```
yarn add react-lazy-io
```

## ⚛️ Components

### Lazy.Container

This components creates a context and uses it to store the newly created
Intersection observer. It takes props as:

- options: Intersection Observer Options

#### Example

```jsx
import React from 'react';
import { Lazy } from 'react-lazy-io';

function Component() {
  return (
    <Lazy.Container>
      {/** React children
            to be lazy loaded can be added here
      */}
    </Lazy.Container>
  );
}
```

### Lazy.Item

This components creates a DOM node and adds its reference to nearest
`Lazy.Container`. The children of this component will be shown depending on the
Container options.

#### Example

```jsx
import React from 'react';
import { Lazy } from 'react-lazy-io';
import imageData from './images.json';

function Component() {
  return (
    <Lazy.Container>
      {imageData.map((data) => (
        <Lazy.Item>
          <img
            key={data.id}
            src={data.url}
            width={data.width}
            height={data.height}
            alt={data.id}
          />
        </Lazy.Item>
      ))}
    </Lazy.Container>
  );
}
```

### Lazy.Placeholder

This components is used as a placeholder in `Lazy.Item` component. The
`Lazy.Item` component uses a slots based mechanism to determine when to show
placeholder and rest of children.

#### Example

```jsx
import React from 'react';
import { Lazy } from 'react-lazy-io';
import imageData from './images.json';

function Component() {
  return (
    <Lazy.Container>
      {imageData.map((data) => (
        <Lazy.Item>
          <Lazy.Placeholder
            className="placeholder"
            style={{ width: `${data.width}px`, height: `${data.height}px` }}
          />
          <img
            key={data.id}
            src={data.url}
            width={data.width}
            height={data.height}
            alt={data.id}
          />
        </Lazy.Item>
      ))}
    </Lazy.Container>
  );
}
```

### Lazy.Image

This components is a sweet lazy loaded image baked with simple transition. The
placeholder for the image is passed on as a child of it, in the form of
`Lazy.Placeholder`. It takes various props as:

- as: DOM node tag name
- height: Image height
- width: Image width
- src: Image src
- alt: Image alt
- transitionDuration: Image visibility transition value in seconds

#### Example

```jsx
import React from 'react';
import { Lazy } from 'react-lazy-io';
import imageData from './images.json';

function Component() {
  return (
    <Lazy.Container
      options={{
        rootMargin: '0px 0px 100px 0px',
      }}
    >
      {imageData.map((data) => (
        <Lazy.Image
          src={data.url}
          width={data.width}
          height={data.height}
          alt={data.id}
          transitionDuration={1.5}
        >
          <Lazy.Placeholder
            className="placeholder"
            style={{ width: `${data.width}px`, height: `${data.height}px` }}
          />
        </Lazy.Image>
      ))}
    </Lazy.Container>
  );
}
```

### Lazy.Suspense

Here I tried to combine Intersection Observer with React Suspense and lazy API
that will load Dynamic component when ever it is about to appear close to
browser viewport. The placeholder for the dynamic component is passed on as a
child of it, in the form of `Lazy.Placeholder` which is also passed as fallback
to Suspense

#### Example

```jsx
import React from 'react';
import { Lazy } from 'react-lazy-io';
const MyAwesomeComponent = React.lazy(() => import('./MyAwesomeComponent'));

function Component() {
  return (
    <Lazy.Container
      options={{
        rootMargin: '0px 0px 100px 0px',
      }}
    >
      <Lazy.Suspense>
        <Lazy.Placeholder className="placeholder" />
        <MyAwesomeComponent />
      </Lazy.Suspense>
    </Lazy.Container>
  );
}
```

## 🪝 Hooks

### useLazyItem

This is a hooks for for `Lazy.Item` component for more granular control of DOM
node. It returns 2 values:

- ref: Reference to the DOM node.
- visibile: If the ref node is visible based on options provided to container.

```jsx
import React from 'react';
import { Lazy, useLazyItem } from 'react-lazy-io';


function CustomItem() {
    const {visible, ref} = useLazyItem();
    return <div ref={ref}>{visible? : "I was visibile": "I'm not visible"}</div>
}

function Component(){
  return (
    <Lazy.Container>
        Testing hooks
        <CustomItem />
    </Lazy.Container>
  );
};
```

## Features

- 💫 Compound Component structure
- 🌳 Tree-shakeable
- 💥 [Tiny bundle](https://bundlephobia.com/package/react-lazy-io) ~1.6 kB
  gzipped
- 🖥 SSR friendly

## 📟 Polyfill

As the library depends on Intersection Observer web API which is almost
supported by all modern browser. But if in case you have some application for
older browser especially <= IE11, then I would suggest to also import a
[intersection-observers](https://www.npmjs.com/package/intersection-observer)
side effect into your project.

```jsx
// import the polyfill before using this library.
import 'intersection-observer';
```

## ✍️ Author

[@imshubhamsingh\_](https://twitter.com/imshubhamsingh_)

P.S. This is my first time creating a fully typed library. I learnt a lot here.
Maybe in coming days will share my learning via a blog.

## 🗂 TODO

- [x] Create Intersection Observer Component and its child components
- [x] Explore `Lazy.Image` component
- [x] SSR friendly
- [x] Explore `Lazy.Suspense` component
- [ ] Add tests - I guess to lazy to write it.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[standard-version](https://github.com/conventional-changelog/standard-version)
for commit guidelines.

## [0.1.0](https://github.com/imshubhamsingh/react-lazy-io/compare/v0.0.2...v0.1.0) (2021-11-16)

### Features

- added lazy suspense
  ([07cc778](https://github.com/imshubhamsingh/react-lazy-io/commit/07cc77876f7d07dd6952176ace51eb3861249287))

### [0.0.2](https://github.com/imshubhamsingh/react-lazy-io/compare/v0.0.1...v0.0.2) (2021-11-15)

### Features

- lazy Image support
  ([4da1c30](https://github.com/imshubhamsingh/react-lazy-io/commit/4da1c30894cd30b7cef50d1080a0a75b7ad8a065))

### 0.0.1 (2021-11-15)

### Features

- added useLazyItem hook support
  ([e45638a](https://github.com/imshubhamsingh/react-lazy-io/commit/e45638a32ebb61ac39c879e06e3c23edd7ae1077))
- basic package working
  ([6828301](https://github.com/imshubhamsingh/react-lazy-io/commit/68283012c9ea34d62e46a5ce610cf7d7c51ca2f1))


src/Lazy.ts
import LazyContainer from './Lazy.Container';
import LazyImage from './Lazy.Image';
import LazyItem, { useLazyItem } from './Lazy.Item';
import LazyPlaceHolder from './Lazy.Placeholder';
import LazySuspense from './Lazy.Suspense';

export const Lazy = {
  /**
   * Lazy Container - It adds single Intersection Observer for Items
   */
  Container: LazyContainer,
  /**
   * Lazy Item - It use nearest Lazy.Container's Intersection Observer to controll the visibility of its children
   */
  Item: LazyItem,
  /**
   * Lazy Placeholder - Anything children inside Placeholder is shown till the time nearest Lazy.Item visibility is false
   */
  Placeholder: LazyPlaceHolder,
  /**
   * Lazy image - A basic lazy loaded transition image by controlling opacity
   */
  Image: LazyImage,
  /**
   * Lazy Suspense - Fetch Components dynamically using Suspense with help of Intersection Observer
   */
  Suspense: LazySuspense,
};

export { useLazyItem };


.eslintrc.js
module.exports = {
  extends: ['plugin:@typescript-eslint/recommended'],
  parser: '@typescript-eslint/parser',
  plugins: [
    'react-hooks',
    'immutable',
    'import',
    '@typescript-eslint',
    'prettier',
  ],
  settings: {
    'import/resolver': {
      node: {
        paths: ['src'],
      },
    },
  },
  rules: {
    'arrow-parens': 0,
    'implicit-arrow-linebreak': 0,
    'no-confusing-arrow': 0, // 1,
    'no-unused-vars': 0, // 2,
    'no-use-before-define': 0,
    'object-curly-newline': 0,
    'operator-linebreak': 1,

    //
    // import rules
    //
    'import/order': 0, // 1,
    // 'import/named': 2,
    'import/no-default-export': 0,
    'import/no-extraneous-dependencies': 1,
    'import/no-unresolved': 0, // 2,
    'import/prefer-default-export': 0,
    //
    // @typescript rules
    //
    '@typescript-eslint/array-type': 0, // [1, 'array-simple'],
    '@typescript-eslint/ban-ts-comment': 1,
    '@typescript-eslint/indent': 0, // ['warn', 2],
    '@typescript-eslint/explicit-member-accessibility': 0, // 2,
    '@typescript-eslint/explicit-module-boundary-types': 0, // 2
    '@typescript-eslint/explicit-function-return-type': 0, // 1,
    '@typescript-eslint/no-explicit-any': 0, // 2,
    '@typescript-eslint/no-object-literal-type-assertion': 0,
    '@typescript-eslint/no-require-imports': 1,
    '@typescript-eslint/no-unused-vars': 0, // 2,
    '@typescript-eslint/no-use-before-define': 0,
    '@typescript-eslint/prefer-interface': 0,

    //
    // immutable rules
    //
    'immutable/no-let': 0, // 1,
    'immutable/no-this': 0, // 1,
    'immutable/no-mutation': 0, // 1,

    //
    // react rules
    //
    // 'react/destructuring-assignment': 1,
    'react/jsx-filename-extension': 0,
    'react/prop-types': 0,
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn',
  },
};


babel.config.js
module.exports = {
  presets: [
    [
      '@babel/preset-react',
      {
        runtime: 'automatic',
      },
    ],
    ['@babel/preset-typescript'],
    [
      '@babel/preset-env',
      {
        useBuiltIns: 'usage',
        corejs: 3,
        loose: true,
        shippedProposals: true,
        targets: {
          browsers: [
            '>5%',
            'last 2 Chrome versions',
            'last 2 Edge versions',
            'last 2 Firefox versions',
            'last 2 Safari versions',
          ],
        },
      },
    ],
  ],
  plugins: [
    ['@babel/plugin-proposal-class-properties', { loose: true }],
    ['@babel/plugin-proposal-private-methods', { loose: true }],
    ['@babel/plugin-proposal-private-property-in-object', { loose: true }],
    '@babel/plugin-transform-runtime',
  ],
};


src/observer.ts
type IGenerateObserver = {
  options: IntersectionObserverInit;
};

export type IObserverCallback = (
  inViewPort: boolean,
  entry: IntersectionObserverEntry,
) => void;

export type INodes = Map<Element, Array<IObserverCallback>>;

type IObserver = {
  callback: IObserverCallback;
  node: Element;
  nodes: INodes;
  observer: IntersectionObserver;
};

export type Observer = {
  nodes: INodes;
  observer: IntersectionObserver;
};

export function generateObserver({ options }: IGenerateObserver): Observer {
  const nodes: INodes = new Map();
  const observer = new IntersectionObserver((entries) => {
    entries.forEach((entry) => {
      const inViewport = entry.isIntersecting;

      nodes.get(entry.target)?.forEach((callback) => {
        callback(inViewport, entry);
      });
    });
  }, options);
  return { nodes, observer };
}

export function observer({ node, nodes, callback, observer }: IObserver) {
  const ogCallbacks = nodes.get(node) || [];
  if (!nodes.has(node)) {
    nodes.set(node, ogCallbacks);
  }

  ogCallbacks.push(callback);

  observer.observe(node);

  return function removeObserver() {
    ogCallbacks.splice(ogCallbacks.indexOf(callback), 1);

    if (ogCallbacks.length === 0) {
      nodes.delete(node);
      observer.unobserve(node);
    }
    if (nodes.size === 0) {
      observer.disconnect();
    }
  };
}


rollup.config.js
import typescript from '@rollup/plugin-typescript';
import resolve from '@rollup/plugin-node-resolve';
import babel from '@rollup/plugin-babel';
import commonjs from '@rollup/plugin-commonjs';
import { terser } from 'rollup-plugin-terser';
import dts from 'rollup-plugin-dts';

import pkg from './package.json';

const isProduction = (env) => env === 'production';

/**
 * @type {import('rollup').RollupOptions}
 */
export default (args) => {
  process.env.NODE_ENV = args.environment;
  console.log(process.env.NODE_ENV);
  let config = {
    input: ['./src/Lazy.ts'],
    output: [
      {
        dir: './',
        format: 'esm',
        entryFileNames: pkg.module,
      },
      {
        dir: './',
        format: 'commonjs',
        entryFileNames: pkg.main,
      },
      {
        dir: './',
        format: 'commonjs',
        entryFileNames: pkg.main,
      },
      {
        dir: './',
        format: 'umd',
        name: pkg.name,
        entryFileNames: pkg.unpkg,
      },
    ],
    external: Object.keys(pkg.peerDependencies),
    plugins: [
      commonjs(),
      typescript(),
      resolve(),
      babel({ exclude: '**/node_modules', babelHelpers: 'runtime' }),
      ...(isProduction(process.env.NODE_ENV)
        ? [
            terser({
              compress: {
                drop_console: true,
                drop_debugger: true,
              },
            }),
          ]
        : []),
    ],
  };

  const types = {
    // path to your declaration files root
    input: './dist/dts/Lazy.d.ts',
    output: [{ file: 'dist/dts/index.d.ts', format: 'es' }],
    plugins: [dts()],
  };

  return [config, types];
};


examples/cra/src/App.js
import React from 'react';
import { Lazy } from 'react-lazy-io';
import './App.css';

const MyAwesomeComponent = React.lazy(() => import('./MyAwesomeComponent'));

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
      <div>
        <h1>Lazy Load Images</h1>
        <Lazy.Container
          options={{
            rootMargin: '0px 0px 0px 0px',
          }}
        >
          <Lazy.Suspense>
            <Lazy.Placeholder className="placeholder" />
            <MyAwesomeComponent />
          </Lazy.Suspense>
        </Lazy.Container>
      </div>
    </div>
  );
}

export default App;


examples/next/pages/index.js
import Head from 'next/head';
import Image from 'next/image';
import styles from '../styles/Home.module.css';
import { Lazy } from 'react-lazy-io';

import imageData from '../data/images.json';

export default function Home() {
  return (
    <div className={styles.container}>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <h1 className={styles.title}>
          Welcome to <a href="https://nextjs.org">Next.js!</a>
        </h1>

        <Lazy.Container
          options={{
            rootMargin: '0px 0px 100px 0px',
          }}
        >
          {imageData.map((data) => (
            <Lazy.Image
              src={data.url}
              width={data.width}
              height={data.height}
              alt={data.id}
              transitionDuration={1.5}
            >
              <Lazy.Placeholder
                className="placeholder"
                style={{ width: `${data.width}px`, height: `${data.height}px` }}
              />
            </Lazy.Image>
          ))}
        </Lazy.Container>
      </main>

      <footer className={styles.footer}>
        <a
          href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
          target="_blank"
          rel="noopener noreferrer"
        >
          Powered by{' '}
          <span className={styles.logo}>
            <Image src="/vercel.svg" alt="Vercel Logo" width={72} height={16} />
          </span>
        </a>
      </footer>
    </div>
  );
}
